src/network/ConnectionRegistry.ts

```diff
@@ -0,0 +1,196 @@
+     */
+    INITIAL = 0,
+
+    /**
+     * When a client switches to a new cluster, it moves to this state.
+     * It means that the client has connected to a new cluster but not sent
+     * its local state to the new cluster yet.
+     */
+    CONNECTED_TO_CLUSTER = 1,
+
+    /**
+     * When a client sends its local state to the cluster it has connected,
+     * it switches to this state. When a client loses all connections to
+     * the current cluster and connects to a new cluster, its state goes
+     * back to {@link CONNECTED_TO_CLUSTER}.
+     * <p>
+     * Invocations are allowed in this state.
+     */
+    INITIALIZED_ON_CLUSTER = 2,
+}
+
+export interface ConnectionRegistry {
+    /**
+     * Returns if the registry active
+     * @return Return true if the registry is active, false otherwise
+     */
+    isActive(): boolean;
+
+    /**
+     * Returns if the registry empty
+     * @return true if the registry empty, false otherwise
+     */
+    isEmpty(): boolean;
+
+    /**
+     * Returns connection by UUID
+     * @param uuid UUID that identifies the connection
+     * @return ClientConnection if there is a connection with the UUID, undefined otherwise
+     */
+    getConnection(uuid: UUID): ClientConnection | undefined;
+
+    /**
+     * Returns all active connections in the registry
+     * @return Array of ClientConnection objects
+     */
+    getConnections(): ClientConnection[];
+
+    /**
+     * Returns a random connection from active connections
+     * @return ClientConnection if there is at least one connection, otherwise null
+     */
+    getRandomConnection(): ClientConnection | null;
+
+    /**
+     * Runs the given function for every active connection
+     * @param fn A function that takes ClientConnection as arg, does not return anything
+     */
+    forEachConnection(fn: (conn: ClientConnection) => void): void;
+
+    /**
+     * Returns if invocation allowed. Invocation is allowed only if client state is {@link INITIALIZED_ON_CLUSTER}
+     * and there is at least one active connection.
+     * @return Error if invocation is not allowed, null otherwise
+     */
+    checkIfInvocationAllowed(): Error | null;
+
+    /**
+     * Delete connection from active connections
+     * @param uuid UUID that identifies the connection
+     */
+    deleteConnection(uuid: UUID): void;
+
+    /**
+     * Returns client state.
+     * @return ClientState enum value
+     */
+    getClientState(): ClientState;
+
+    /**
+     * Sets the client state
+     * @param clientState
+     */
+    setClientState(clientState: ClientState): void;
+
+    /**
+     * Adds or updates a client connection by uuid
+     * @param uuid UUID to identify the connection
+     * @param connection the ClientConnection to set
+     */
+    setConnection(uuid: UUID, connection: ClientConnection): void
+}
+
+export class ConnectionRegistryImpl implements ConnectionRegistry {
+
+    private active = false;
```

Comment:
This flag should be mutated to `true` when the connection manager starts and set to `false` on shutdown.

Reply:
So need for `alive` then?

Reply:
Plus, we can remove altogether, and pass and use connection registry instead of connection manager

Reply:
> pass and use connection registry instead of connection manager

This sounds like a good option to me.